R/result.R

```diff
@@ -1,36 +1,285 @@
+      out$output <- x$output
+
+      # append metadata fields to dataframe as attributes
+      metadata_elements <- c("name", "tags", "info", "context", "github", "options")
+      for (element in metadata_elements) {
+        if (element %in% names(x)) {
+          attr(out, element) <- x[[element]]
+        }
+      }
+
+      out
+    }
+  ),
+
+  active = list(
+    name = function(name) private$get_or_set(variable = ".name", value = name),
+    result = function(result) private$get_or_set(variable = ".result", value = result),
+    params = function(params) private$get_or_set(variable = ".params", value = params),
+    tags = function(tags) private$get_or_set(variable = ".tags", value = tags),
+    info = function(info) private$get_or_set(variable = ".info", value = info),
+    context = function(context) private$get_or_set(variable = ".context", value = context),
+    github = function(github) private$get_or_set(variable = ".github", value = github),
+    options = function(options) private$get_or_set(variable = ".options", value = options),
+    output = function(output) private$get_or_set(variable = ".output", value = output),
+    rscript = function(rscript) private$get_or_set(variable = ".rscript", value = rscript),
+
+    params_summary = function() {
+      d <- self$params
+
+      d$packages <- NULL
+
+      to_list_col <- lengths(d) != 1
+      d[to_list_col] <- lapply(d[to_list_col], list)
+
+      d <- dplyr::as_tibble(d)
+      d$did_error <- FALSE
+      d
+    }
+  ),
+
+  private = list(
+    .name = NULL,
+    .result = NULL,
+    .params = NULL,
+    .tags = NULL,
+    .info = NULL,
+    .context = NULL,
+    .github = NULL,
+    .options = NULL,
+    .output = NULL,
+    .rscript = NULL
+  )
+)
+
+
+BenchmarkFailure <- R6.1Class(
+  classname = "BenchmarkFailure",
+  inherit = Serializable,
+
+  public = list(
+    initialize = function(error,
+                          params) {
+      self$error <- error
+      self$params <- params
+    }
+  ),
+  active = list(
+    error = function(error) private$get_or_set(variable = ".error", value = error),
+    params = function(params) private$get_or_set(variable = ".params", value = params),
+
+    params_summary = function() {
+      d <- self$params
+
+      d$packages <- NULL
+      to_list_col <- lengths(d) != 1
+      d[to_list_col] <- lapply(d[to_list_col], list)
+
+      d <- dplyr::as_tibble(d)
+      d$did_error <- TRUE
+      d
+    }
+  ),
+  private = list(
+    .error = NULL,
+    .params = NULL
+  )
+)
+
+
+BenchmarkResults <- R6.1Class(
+  classname = "BenchmarkResults",
+  inherit = Serializable,
+
+  public = list(
+    initialize = function(results) {
+      self$results <- results
+    },
+    data.frame = function(row.names = NULL, optional = FALSE, ...) {
+      x <- self$results
+      valid <- map_lgl(x, ~inherits(.x, "BenchmarkResult"))  # failures will be BenchmarkFailure
+
+      dplyr::bind_rows(lapply(x[valid], function(res) res$data.frame(...)))
```

Comment:
Is there a reason not to make this a tibble? We use data.frames and tibbles interchangeably. Should we standardize?

Reply:
It is because `BenchmarkResult$new(...)$data.frame(...)` returns a tibble. I just named the method `$data.frame()` so it's parallel with the `BenchmarkResult$new(...)$list` and `$json` active bindings. But really it's a method, not an active binding, so maybe it should be called `$to_dataframe()` or something? Also `data.frame` vs. `dataframe` vs. `data_frame` vs. `data frame` vs. `tibble` is a mess...I don't know what the most intuitive is from a user perspective.

tl;dr: kind of want to change this name; no idea what to change it to; suggestions welcome

Reply:
Ah I see. Maybe `to_tibble` or `bm_tibble` might be good too.

Reply:
Yeah, I'm leaning towards `$to_dataframe()` or `$to_tibble()`